### PR TITLE
[Tidy] Move figure layout dashboard overrides into a template

### PR DESCRIPTION
**Pull request by antonymilne**
## Description

Completely closes 
https://github.com/McK-Internal/vizro-internal/issues/1208. This should now be the last thing we need to do on graph templates 🥳 

Now that templates are moving entirely to the clientside there's a much cleaner way to apply the figure layout updates:
* better to apply them through a template rather than `fig.layout` so there's no need to do the `if fig.layout.property is None` check. 
* `vizro_light` and `vizro_dark` are unchanged and still optimised for use outside dashboards. Now we merge in a new theme `dashboard_template_overrides` when they're applied in the dashboard. User changes to `vizro_light/dark` are still maintained since we just merge the extra `dashboard_template_overrides` on top of them. The new `dashboard_template_overrides` is _not_ public and so not registered in `pio.templates`
* the `vizro_light/dark + dashboard_template_overrides` is optimised for charts with no titles, so that all the weird custom logic that alters `fig.layout` now only applies when a `title` is set. This will probably be removed soon.

**This is just refactoring. Nothing should have changed here either inside or outside dashboard.**

@huong-li-nguyen please could you double check I didn't change anything accidentally here? As I understand it, this should be the behaviour both before and after this PR:

**Standalone chart**

```
title_pad_l = 24
title_pad_r = 24
title_pad_t = 24
title_pad_b = 0

margin_l = 80
margin_r = 24
margin_t = 64
margin_b = 64
```

**Chart in dashboard with no title**

Same as standalone chart except `title_pad_l, title_pad_r, margin_l, margin_t` different:

```
title_pad_l = 0
title_pad_r = 0
title_pad_t = 24
title_pad_b = 0

margin_l = 24
margin_r = 24
margin_t = 24
margin_b = 64
```

**Chart in dashboard with title and no subtitle**

Same as _chart in dashboard with no title_ except `title_pad_t = 7`.

**Chart in dashboard with title and no subtitle**

Same as _chart in dashboard with no title_ except `margin_t = 64` (what it was in standalone chart)

## Screenshot

image

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

  - I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
  - I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorized to submit this contribution on behalf of the original creator(s) or their licensees.
  - I certify that the use of this contribution as authorized by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
  - I have not referenced individuals, products or companies in any commits, directly or indirectly.
  - I have not added data or restricted code in any commits, directly or indirectly.

### CODE REVIEW

File: vizro-core/src/vizro/models/_components/graph.py [hunk 1]
```diff
@@ -121,22 +122,14 @@ def _filter_interaction(
...
-
-        if fig.layout.title.pad.l is None:
-            fig.update_layout(title_pad_l=0)
-
-        if fig.layout.title.pad.r is None:
-            fig.update_layout(title_pad_r=0)
-
-        if fig.layout.margin.l is None:
-            fig.update_layout(margin_l=24)
+        if fig.layout.title.text:
```
Comment (antonymilne) [inline vizro-core/src/vizro/models/_components/graph.py:125]: @huong-li-nguyen the idea here is that I've made the `dashboard_overrides` optimised for a chart with no title so `margin_t=24` there. Hence the logic here now just tweaks things when we have a title.

In future when we have `header` I guess we will just remove this whole `_optimise` method?

File: vizro-core/tests/unit/vizro/models/_components/test_graph.py [hunk 2]
```diff
@@ -99,7 +99,7 @@ def test_getitem_unknown_args(self, standard_px_chart):
 
     @pytest.mark.parametrize(
```
Comment (antonymilne) [inline vizro-core/tests/unit/vizro/models/_components/test_graph.py:100]: These test cases should render the same way as before, it's just that the way we apply the values is different now.

File: vizro-core/src/vizro/models/_components/graph.py (hunk 1 above)
Comment (antonymilne) [reply vizro-core/src/vizro/models/_components/graph.py:125]: If you think this is a bad way to do it (e.g. if we should keep `margin_t=64` and not override it in `dashboard_overrides`) then please do say 🙂

Comment (huong-li-nguyen) [reply vizro-core/src/vizro/models/_components/graph.py:125]: No, that works for me! That they have a title is probably the rarer case, so this way around also works (and will be called less)!

File: vizro-core/tests/unit/vizro/actions/conftest.py [hunk 3]
```diff
@@ -40,14 +40,12 @@ def box_params():
 
 @pytest.fixture
 def box_chart(gapminder_2007, box_params):
-    return px.box(gapminder_2007, **box_params).update_layout(margin_t=24, title_pad_l=0, title_pad_r=0, margin_l=24)
+    return px.box(gapminder_2007, **box_params)
```
Comment (huong-li-nguyen) [inline vizro-core/tests/unit/vizro/actions/conftest.py:43]: Ahh, very satisfying to see this go 🥳

**Review by huong-li-nguyen — APPROVED** [review]
This looks great! 🚀  Very satisfying to see some of the code go 😄 

It's all good from my side. I didn't manually test anything, but if the unit tests and screenshots tests have passed + the scratch example looks fine, it should be all good!

File: vizro-core/src/vizro/models/_components/graph.py [hunk 4]
```diff
@@ -49,6 +49,7 @@ class Graph(VizroBaseModel):
     _set_actions = _action_validator_factory("clickData")
     _validate_callable = validator("figure", allow_reuse=True)(_process_callable_data_frame)
 
+    # TODO: change to using template rather than layout. Test overridding works as expected.
```
Comment (antonymilne) [inline vizro-core/src/vizro/models/_components/graph.py:52]:
```suggestion
```

**Review by petar-qb — APPROVED** [review]
Thanks for the code and the tests refactoring. ⚡ It will be much easier to maintain themes with these changes.  

 
Here's a screenshot of the scratch dev app. Are we satisfied with the current margins and padding? The reason I ask is that the titles of the graphs, those with subtitles, seem to be positioned too far from the top border of their container. Do you think we should adjust this spacing?
image

**Comment by antonymilne** [comment]
Great observation @petar-qb, thanks for the careful testing 👍 

My guess this is because the `title_pad_t=7` that we apply when there's a subtitle is just a rough estimate and it's impossible to get it better than this. But @huong-li-nguyen will know better:
1. has the alignment got worse here due to my changes or did it always look like this?
2. should we try and tweak `title_pad_t` to align better?

Unless @huong-li-nguyen thinks otherwise then I'd say we just leave it how it is because we're going to be removing the whole `optimise_fig_layout_for_dashboard` soon in favour of new fields `Graph.title`, `Graph.header`, `Graph.footer` (and similarly for other models) which will be the new, improved way of doing titles and subtitles. When that change is made we won't have any alignment issues any more and we won't need to hack around with `title_pad_t` at all.

**Comment by huong-li-nguyen** [comment]
> Great observation @petar-qb, thanks for the careful testing 👍
> 
> My guess this is because the `title_pad_t=7` that we apply when there's a subtitle is just a rough estimate and it's impossible to get it better than this. But @huong-li-nguyen will know better:
> 
> 1. has the alignment got worse here due to my changes or did it always look like this?
> 2. should we try and tweak `title_pad_t` to align better?
> 
> Unless @huong-li-nguyen thinks otherwise then I'd say we just leave it how it is because we're going to be removing the whole `optimise_fig_layout_for_dashboard` soon in favour of new fields `Graph.title`, `Graph.header`, `Graph.footer` (and similarly for other models) which will be the new, improved way of doing titles and subtitles. When that change is made we won't have any alignment issues any more and we won't need to hack around with `title_pad_t` at all.

Just leave it as is. It's really icky to get this right, because the graph.title is coming from Plotly so the normal margin / padding specifications don't apply as if this were a completely new html container 😅  I don't think many people use subtitles anyway and as @antonymilne says, I'll change that very soon anyway.

**Review by maxschulz-COL — APPROVED** [review]

**Pull request merged**